### DISCUSSION TITLE
ltree datatype support

### DISCUSSION THREAD

**Discussion by mlesin** [elephantry/elephantry#21] [Q&A] [answered]
Is there any support for ltree datatype or it should be implemented as a custom one?
By the way, I can't see somewhere in the docs how to implement custom datatypes, is it hard or I just have to implement FromSql/ToSql traits?

**Comment by sanpii** [comment, marked as answer]
ltree is not supported, you can create a custom type. I thought about it but I didn’t find an appropriate rust type.

The most difficult part to implement FromSql/ToSql is to understand in/out and send/recv functions: https://github.com/postgres/postgres/blob/REL_14_STABLE/contrib/ltree/ltree_io.c

in/out is for text format, easily to undertand but require a parser.
send/recv is for binary format, more difficult to understand but easier to convert data in rust.

**Comment by sanpii** [comment]
I made it quickly by wrapping a `ltree` in a `Vec<String>` : https://github.com/elephantry/elephantry/blob/ltree/core/src/sql/ltree.rs (in the ltree branch).

Let me know if that match your need. Is it necessary to also add a type for `lquery` and `ltxtquery`?

**Reply by mlesin** [reply]
Thank you so much!
That is also great example of implementing custom type!
Yes, usually at least lquery is needed when working with ltree datatype, if both will be implemented it would be perfect match.
I have one doubt hovewer, why not to store it internally as a regular String? In that case it will require only one memory allocation instead of 1+N elements in path, and then one more during serialization. Probably it would be more effective that way?
Also, Ltree labels have a restriction on contents, according to documentation, each label shoud match `A-Za-z0-9_+` regex and be not more than 255 characters long. Maybe checking this invariant during string conversion is not a bad idea. But it also add some cost on the other hand.

**Reply by sanpii** [reply]
I added `lquery` and `ltxtquery`.

>  have one doubt hovewer, why not to store it internally as a regular String?

To easily manipulate ltree in rust with `Vec::contains`, `Vec::push` or `Vec::pop` for example.

> Also, Ltree labels have a restriction on contents,

I don’t really think about making checks in the rust side, but I think it’s not a good idea:
- postgresql already do that: less code, less bugs;
- performance: for ltree, checking each label with a regex cost many time;
- it’s not necessarily exhaustive: the user can add more constraints on a type;
- too restrictive: I might want to create a new type, without this restriction. In this case I can’t reuse the code.

And I don’t find positive point to do that. Better error handling?
